spec/ics-4-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,384 @@
+}
+```
+
+```typescript
+function chanOpenAck() {
+  channel = get("connections/{connectionIdentifier}/channels/{channelIdentifier}")
+  assert(channel.state === INIT)
+  assert(getCallingModule() === channel.moduleIdentifier)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}")
+  assert(verifyMembership(
+    consensusState.getRoot(),
+    proofTry,
+    "connections/{connection.counterpartyConnectionIdentifier}/channels/{channel.counterpartyChannelIdentifier}",
+    Channel{TRYOPEN, channel.counterpartyModuleIdentifier, channel.moduleIdentifier, channelIdentifier, 0, 0}
+  ))
+  channel.state = OPEN
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
+}
+```
+
+```typescript
+interface ChanOpenConfirm {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  proofAck: CommitmentProof
+}
+```
+
+```typescript
+function chanOpenConfirm() {
+  channel = get("connections/{connectionIdentifier}/channels/{channelIdentifier}")
+  assert(channel.state === TRYOPEN)
+  assert(getCallingModule() === channel.moduleIdentifier)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}")
+  assert(verifyMembership(
+    consensusState.getRoot(),
+    proofAck,
+    "connections/{connection.counterpartyConnectionIdentifier}/channels/{channel.counterpartyChannelIdentifier}",
+    Channel{OPEN, channel.counterpartyModuleIdentifier, channel.moduleIdentifier, channelIdentifier, 0, 0}
+  ))
+  channel.state = OPEN
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
+}
+```
+
+### Channel closing handshake
+
+(todo, isomorphic to connection closing handshake)
+
+![packet-state-machine](packet-state-machine.png)
+
+### Sending packets
+
+`sendPacket` is called by a module.
+
+```typescript
+function sendPacket(packet: Packet) {
+  channel = get("connections/{packet.sourceConnection}/channels/{packet.sourceChannel}")
+  assert(channel.state === OPEN)
+  assert(getCallingModule() === channel.moduleIdentifier)
+  assert(packet.destChannel === channel.counterpartyChannelIdentifier)
+  connection = get("connections/{packet.sourceConnection}")
+  assert(connection.state === OPEN)
+  assert(packet.destConnection === connection.counterpartyConnectionIdentifier)
+  consensusState = get("clients/{connection.clientIdentifier}")
+  assert(consensusState.getHeight() < packet.timeoutHeight)
+  assert(packet.sequence === channel.nextSequenceSend)
+  channel.nextSequenceSend = channel.nextSequenceSend + 1
+  set("connections/{packet.sourceConnection}/channels/{packet.sourceChannel}", channel)
+  set("connections/{packet.sourceConnection}/channels/{packet.sourceChannel}/packets/{sequence}", commit(packet.data))
+}
+```
+
+### Receiving packets
+
+`recvPacket` is called by a module.
```

Comment:
Maybe call it "receivedPacket()" or "receivePacket()". What causes a module to call this?

Reply:
Hmm, I think this tense is appropriate. Added an explanation.